app/code/Magento/InventoryLowQuantityNotification/Plugin/InventoryApi/SourceItemsDeleteInterfacePlugin.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryLowQuantityNotification\Plugin\InventoryApi;
+
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryApi\Api\SourceItemsDeleteInterface;
+use Magento\InventoryLowQuantityNotification\Model\ResourceModel\SourceItemConfiguration\Delete;
+
+/**
+ * This plugin keeps consistency between SourceItem and SourceItemConfiguration while deleting
+ */
+class SourceItemsDeleteInterfacePlugin
+{
+    /**
+     * @var Delete
+     */
+    private $delete;
+
+    /**
+     * @param Delete $delete
+     */
+    public function __construct(
+        Delete $delete
+    ) {
+        $this->delete = $delete;
+    }
+
+    /**
+     * Keep database consistency while a source item is removed
+     *
+     * @param SourceItemsDeleteInterface $subject
+     * @param void $result
+     * @param array $sourceItems
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterExecute(
+        SourceItemsDeleteInterface $subject,
+        $result,
+        array $sourceItems
+    ) {
+        foreach ($sourceItems as $sourceItem) {
+            /** @var SourceItemInterface $sourceItem */
+            $this->delete->execute($sourceItem->getSourceCode(), $sourceItem->getSku());
+        }
```

Comment:
a little bit worrying about potential performance degradation. 
As we are pluginizaing Bulk operation with logic which process records one by one, 
this is current issue in LowStock Design, but here should be another bulk operation being called

Reply:
Hello @maghamed ,
actually this plugin is for `Magento\InventoryApi\Api\SourceItemsDeleteInterface` what is not involved in bulk operations. 

The plugin handling the same operation at bulk level is: `Magento\InventoryLowQuantityNotification\Plugin\InventoryCatalogApi\BulkSourceUnassignInterfacePlugin`.

Reply:
Just noticed that I did not answer in this thread and we still have this PR opened.

Looks like we have always some confusion regarding the terminology, when I say `Bulk` API I mean the API which accepts a bunch of entities (an array), but not a single entity like most of the Service Contracts we have (i.e. all Repositories). But looks like you consider `Bulk API` term as an async execution feature.
 
For example, 
ProductRepository::save is a non-bulk API , as it can handle just one product entity per a call.
```
interface ProductRepositoryInterface
{
    /**
     * Create product
     *
     * @param \Magento\Catalog\Api\Data\ProductInterface $product
     * @param bool $saveOptions
     * @return \Magento\Catalog\Api\Data\ProductInterface
     * @throws \Magento\Framework\Exception\InputException
     * @throws \Magento\Framework\Exception\StateException
     * @throws \Magento\Framework\Exception\CouldNotSaveException
     */
    public function save(\Magento\Catalog\Api\Data\ProductInterface $product, $saveOptions = false);
```
SourceItemsDeleteInterface - is a bulk API, as it accepts an array of sourceItems
```
interface SourceItemsDeleteInterface
{
    /**
     * Delete Multiple Source item data
     *
     * @param \Magento\InventoryApi\Api\Data\SourceItemInterface[] $sourceItems
     * @return void
     * @throws \Magento\Framework\Exception\InputException
     * @throws \Magento\Framework\Exception\CouldNotSaveException
     */
    public function execute(array $sourceItems): void;
}

```

So, in this particular situation we have a service 
`SourceItemsDeleteInterface ::execute(array $sourceItems)`
which accepts an array of entities to work with. 

But in the plugin which handles this logic we have an iteration 

```
        foreach ($sourceItems as $sourceItem) {
            /** @var SourceItemInterface $sourceItem */
            $this->delete->execute($sourceItem->getSourceCode(), $sourceItem->getSku());
        }
```
calling ONE api call for each single entity from provided input array. 
which could be quite time consuming if the array is big enough.


Moreover you are using private implementation
`Magento\InventoryLowQuantityNotification\Model\ResourceModel\SourceItemConfiguration\Delete`
, but not the public API from InventoryLowQuantityNotificationApi module
`\Magento\InventoryLowQuantityNotificationApi\Api\DeleteSourceItemConfigurationInterface`
```
namespace Magento\InventoryLowQuantityNotificationApi\Api;

/**
 * Delete the source item configuration
 *
 * @api
 */
interface DeleteSourceItemConfigurationInterface
{
    /**
     * @param string $sourceCode
     * @param string $sku
     * @return void
     */
    public function execute(string $sourceCode, string $sku): void;
}

```

The interesting thing is that there is Bulk Service in the LowStockQuantity module 
`\Magento\InventoryLowQuantityNotification\Model\ResourceModel\BulkConfigurationUnassign::execute`
but it's not public API of the module.

Reply:
Ok, clear now

Reply:
Ok, clear now. I'm reworking it.

Answering to the missing public API for bulk ops:
`\Magento\InventoryLowQuantityNotification\Model\ResourceModel\BulkConfigurationAssign`, `\Magento\InventoryLowQuantityNotification\Model\ResourceModel\BulkConfigurationUnassign` and `\Magento\InventoryLowQuantityNotification\Model\ResourceModel\BulkConfigurationTransfer` do not have an API by intention because they are supposed not to be called directly. The idea is to use them only while operating at source level.

They should only be called by the following plugins to keep database consistency:
- `\Magento\InventoryLowQuantityNotification\Plugin\InventoryCatalogApi\BulkConfigurationTransferInterfacePlugin`
- `\Magento\InventoryLowQuantityNotification\Plugin\InventoryCatalogApi\BulkSourceAssignInterfacePlugin`
- `\Magento\InventoryLowQuantityNotification\Plugin\InventoryCatalogApi\BulkSourceUnassignInterfacePlugin`

I did not want to include the logic in the plugins because I think that plugins should be maintaining too much logic. What do you think?